src/Microsoft.AspNetCore.Mvc.Razor/ViewComponentTagHelpers/ViewComponentTagHelperDescriptorFactory.cs

```diff
@@ -0,0 +1,160 @@
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.Razor.Host;
+using Microsoft.AspNetCore.Mvc.ViewComponents;
+using Microsoft.AspNetCore.Razor.Compilation.TagHelpers;
+using Microsoft.AspNetCore.Razor.Runtime.TagHelpers;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.ViewComponentTagHelpers
+{
+    /// <summary>
+    /// Provides methods to create tag helper representations of view components.
+    /// </summary>
+    public class ViewComponentTagHelperDescriptorFactory
+    {
+        private readonly IViewComponentDescriptorProvider _descriptorProvider;
+
+        /// <summary>
+        /// Creates a new ViewComponentTagHelperDescriptorFactory than creates tag helper descriptors for
+        /// view components in the given descriptorProvider.
+        /// </summary>
+        /// <param name="descriptorProvider">The provider of view component descriptors.</param>
+        public ViewComponentTagHelperDescriptorFactory(IViewComponentDescriptorProvider descriptorProvider)
+        {
+            _descriptorProvider = descriptorProvider;
+        }
+
+        /// <summary>
+        /// Creates <see cref="TagHelperDescriptor"/> representations of view components in a given assembly.
+        /// </summary>
+        /// <param name="assemblyName">The name of the assembly containing the view components to translate.</param>
+        /// <returns>A <see cref="IEnumerable{TagHelperDescriptor}"/>, one for each view component.</returns>
+        public IEnumerable<TagHelperDescriptor> CreateDescriptors(string assemblyName)
+        {
+            if (assemblyName == null)
+            {
+                throw new ArgumentNullException(assemblyName);
+            }
+
+            var viewComponentDescriptors = _descriptorProvider.GetViewComponents()
+                .Where(viewComponent => assemblyName.Equals(
+                viewComponent.TypeInfo.Assembly.GetName().Name));
+
+            return CreateDescriptors(viewComponentDescriptors);
+        }
+
+        private IEnumerable<TagHelperDescriptor> CreateDescriptors(
+            IEnumerable<ViewComponentDescriptor> viewComponentDescriptors)
+        {
+            var tagHelperDescriptors = new List<TagHelperDescriptor>();
+
+            foreach (var viewComponentDescriptor in viewComponentDescriptors)
+            {
+                var tagHelperDescriptor = CreateDescriptor(viewComponentDescriptor);
+                tagHelperDescriptors.Add(tagHelperDescriptor);
+            }
+
+            return tagHelperDescriptors;
+        }
+
+        private TagHelperDescriptor CreateDescriptor(ViewComponentDescriptor viewComponentDescriptor)
+        {
+            // Fill in the attribute and required attribute descriptors.
+            IEnumerable<TagHelperAttributeDescriptor> attributeDescriptors;
+            IEnumerable<TagHelperRequiredAttributeDescriptor> requiredAttributeDescriptors;
+            if (!TryGetAttributeDescriptors(viewComponentDescriptor,
```

Comment:
Since we don't have validation yet you should probably restructure this method to start with a TagHelperDescriptor and then call into builder methods that fill in appropriate properties. That way we can avoid saving up all the information for the descriptor which causes us to do ugly out params :smile: